src/inc/til/generational.h

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+namespace til // Terminal Implementation Library. Also: "Today I Learned"
+{
+    struct generation_t
+    {
+        auto operator<=>(const generation_t&) const = default;
+
+        constexpr void bump() noexcept
+        {
+            _value++;
+        }
+
+        uint32_t _value = 0;
+    };
+
+    template<typename T>
+    struct generational
+    {
+        generational() = default;
+        explicit constexpr generational(auto&&... args) :
+            _value{ std::forward<decltype(args)>(args)... } {}
+        explicit constexpr generational(generation_t generation, auto&&... args) :
+            _generation{ generation },
+            _value{ std::forward<decltype(args)>(args)... } {}
+
+        constexpr bool operator==(const generational& rhs) const noexcept { return generation() == rhs.generation(); }
+        constexpr bool operator!=(const generational& rhs) const noexcept { return generation() != rhs.generation(); }
+
+        constexpr generation_t generation() const noexcept
+        {
+            return _generation;
+        }
+
+        [[nodiscard]] constexpr const T* operator->() const noexcept
+        {
+            return &_value;
+        }
+
+        [[nodiscard]] constexpr const T& operator*() const noexcept

```

Comment:
it seems weird to my brain that these return different types, but I'm sure this is just how c++ is under the hood and I never really thought about it

Reply:
Yeah, this is exactly how `std::unique_ptr` or `std::optional` work. `operator*` is used for `*obj` statements, where you get a reference to the underlying object. `operator->` is used for things like `obj->func()`, where the language expects you to return a pointer from the operator, because `->` is for dereferencing pointers when coming from C.